radix-transactions/src/manifest/manifest_instructions.rs

```diff
@@ -279,34 +279,338 @@ impl ManifestInstruction for AssertWorktopContainsNonFungibles {
-        Effect::WorktopAssertion {
-            assertion: WorktopAssertion::IsEmpty,
+        Effect::ResourceAssertion {
+            assertion: ResourceAssertion::BucketContents {
+                bucket: self.bucket_id,
+                constraint: &self.constraint,
+            },
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Default, ManifestSbor, ScryptoDescribe)]
+#[sbor(transparent)]
+pub struct ManifestResourceConstraints {
+    specified_resources: IndexMap<ResourceAddress, ManifestResourceConstraint>,
+}
+
+impl ManifestResourceConstraints {
+    pub fn specified_resources(&self) -> &IndexMap<ResourceAddress, ManifestResourceConstraint> {
+        &self.specified_resources
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = (&ResourceAddress, &ManifestResourceConstraint)> {
+        self.specified_resources.iter()
+    }
+
+    pub fn is_valid(&self) -> bool {
+        for (resource_address, constraint) in self.iter() {
+            if !constraint.is_valid_for(resource_address) {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+impl IntoIterator for ManifestResourceConstraints {
+    type Item = (ResourceAddress, ManifestResourceConstraint);
+    type IntoIter =
+        <IndexMap<ResourceAddress, ManifestResourceConstraint> as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.specified_resources.into_iter()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, ManifestSbor, ScryptoDescribe)]
+pub enum ManifestResourceConstraint {
+    NonZeroAmount,
+    ExactAmount(Decimal),
+    AtLeastAmount(Decimal),
+    ExactNonFungibles(IndexSet<NonFungibleLocalId>),
+    AtLeastNonFungibles(IndexSet<NonFungibleLocalId>),
+    General(GeneralResourceConstraint),
+}
+
+impl ManifestResourceConstraint {
+    pub fn is_valid_for(&self, resource_address: &ResourceAddress) -> bool {
+        if resource_address.is_fungible() {
+            self.is_valid_for_fungible_use()
+        } else {
+            self.is_valid_for_non_fungible_use()
+        }
+    }
+
+    pub fn is_valid_for_fungible_use(&self) -> bool {
+        match self {
+            ManifestResourceConstraint::NonZeroAmount => true,
+            ManifestResourceConstraint::ExactAmount(amount) => !amount.is_negative(),
+            ManifestResourceConstraint::AtLeastAmount(amount) => !amount.is_negative(),
+            ManifestResourceConstraint::ExactNonFungibles(_) => false,
+            ManifestResourceConstraint::AtLeastNonFungibles(_) => false,
+            ManifestResourceConstraint::General(general) => general.is_valid_for_fungible_use(),
+        }
+    }
+
+    pub fn is_valid_for_non_fungible_use(&self) -> bool {
+        match self {
+            ManifestResourceConstraint::NonZeroAmount => true,
+            ManifestResourceConstraint::ExactAmount(amount) => !amount.is_negative(),
+            ManifestResourceConstraint::AtLeastAmount(amount) => !amount.is_negative(),
```

Comment:
Minor: do we want to restrict granularity to 1?

Reply:
Nice idea, have added these constraints.